File: python/demo_for_direc.py
#!/usr/bin/env python
"""
Created on Wed Oct 15 09:28:58 2014

Authors:
1)
Oren Freifeld
Email: freifeld@csail.mit.edu

2)
Yixin Li
Email: liyixin@mit.edu

Example usage of the superpixels code on all files in a directory
"""
import os
from sys import path
if '.'+os.sep not in path:
    path.insert(0,'.'+os.sep)
    
from scipy import misc
import scipy.io as sio

from superpixels.SuperpixelsWrapper import SuperpixelsWrapper
from of.utils import *


def get_list_of_all_imgs_dir(dirname,
                             exts = ['.jpg','.jpeg','.png','.bmp','.ppm']):
    L = os.listdir(dirname) 
    exts = exts + [ item.upper()  for item in exts]     
    L = [ x for x in L if os.path.splitext(x)[-1] in exts]     
    L.sort()    
    return L
    
def main(image_direc = 'image',
        nPixels_on_side = 15,
        i_std = 15, # std dev for color Gaussian
        imgs_of_the_same_size=None
        ):
    if image_direc is None:
        raise ValueError("image_direc cannot be None")
    image_direc = os.path.expanduser(image_direc)
    image_direc = os.path.abspath(image_direc)
    FilesDirs.raise_if_dir_does_not_exist(image_direc)
    imgs = get_list_of_all_imgs_dir(image_direc)     
    if not imgs:
        raise Exception("\n\nDirectory \n{}\ncontains no images.\n".format(image_direc))
    
    # Results will be saved in the /results folder under input directory. You can change it.
    save_path_root = os.path.join(image_direc , 'result')
    print "I am going to save results into",save_path_root
    FilesDirs.mkdirs_if_needed(save_path_root)
    
    # Part 1: Specify the parameters:
    prior_count_const = 5  # the weight of Inverse-Wishart prior of space covariance(ex:1,5,10)
    use_hex = True # toggle between hexagons and squares (for the init)
    prior_weight = 0.5 # in the segmentation, 
                       # we do argmax w * log_prior + (1-w) *log_likelihood.
                       # Keeping w (i.e., prior_weight) at 0.5 means we are trying 
                       # to maximize the true posterior. 
                       # We keep the paramter here in case the user will like 
                       # to tweak it. 
                        
    calc_s_cov = True # If this is False, then we avoid estimating the spatial cov.
    num_EM_iters = nPixels_on_side
    num_inner_iters = 10
    sp_size = nPixels_on_side*nPixels_on_side 

    
    
    
    
    

    for i,img_filename in enumerate(imgs): 

        # Part 2 : prepare for segmentation
        print img_filename
        fullfilename = os.path.join(image_direc,img_filename)
        FilesDirs.raise_if_file_does_not_exist(fullfilename)
        img = misc.imread(fullfilename)
        
        dimx=img.shape[1]
        dimy=img.shape[0]
        
        if not((img.ndim in [1,3]) and (img.shape[2] in [1,3])):
            raise ValueError("\n\nProblem with {0}\n\nI was expecting 3 channels, but img.shape = {1}\n\n".format(fullfilename,img.shape))
        
        tic = time.clock()
        if (imgs_of_the_same_size and i !=0):
            sw.initialize_seg()
            # you can use the same SuperpixelsWrapper object with different imgs and/or, 
            # i_std, s_std, prior_count. 
            # Just call sw.set_img(new_img), sw.initialize_seg(), and/or
            # sw.set_superpixels(i_std=..., s_std = ..., prior_count = ...)
            # again and recompute the seg.
        else:
            sw = SuperpixelsWrapper(dimy=dimy,dimx=dimx, nPixels_in_square_side=nPixels_on_side,
                                 i_std = i_std , s_std = nPixels_on_side, 
                                 prior_count = prior_count_const*sp_size,
                                 use_hex = use_hex 
                                )      
        
        toc = time.clock()
        print 'init time = ', toc-tic
        print 'nSuperpixels =', sw.nSuperpixels    
        sw.set_img(img) 



        # Part 3: Do the superpixel segmentation

        tic  = time.clock() 
        #actual work
        sw.calc_seg(nEMIters=num_EM_iters, nItersInner=num_inner_iters, calc_s_cov=calc_s_cov, prior_weight=prior_weight)
        # Copy the parameters from gpu to cpu
        sw.gpu2cpu() 
        toc  = time.clock()
        print 'superpixel calculation time = ',toc-tic


        # Part 4: Save the mean/boundary image and the resulting parameters

        img_overlaid = sw.get_img_overlaid()   # get the boundary image                      
        img_cartoon = sw.get_cartoon()         # get the cartoon image
        grid = ['square','hex'][sw.use_hex]

        root_slash_img_num = os.path.splitext(img_filename)[0]
        img_num = os.path.split(root_slash_img_num)[1]
        #fname_res_border = '_'.join([img_num , 'std', str(i_std), 'border', grid+'.png'])
        #fname_res_cartoon = '_'.join([img_num , 'std', str(i_std), 'mean', grid+'.png'])
        
        fname_res_border =  '_'.join([img_num,'n','{0:03}'.format(nPixels_on_side), 'std', '{0:02}'.format(i_std), 'border', grid+'.png'])
        fname_res_cartoon = '_'.join([img_num, 'n','{0:03}'.format(nPixels_on_side), 'std', '{0:02}'.format(i_std), 'mean', grid+'.png'])
        
        
        fname_res_border = os.path.join(save_path_root,fname_res_border)
        fname_res_cartoon = os.path.join(save_path_root,fname_res_cartoon)

        print 'saving',fname_res_border
        misc.imsave(fname_res_border, img_overlaid)
        print 'saving',fname_res_cartoon
        misc.imsave(fname_res_cartoon, img_cartoon)

        #save the resulting parameters to MATLAB .mat file 
        #mat_filename = os.path.join(save_path_root , img_num + '_std_'+str(i_std)+'.mat')
        mat_filename = os.path.join(save_path_root#, img_num + '_std_'+str(i_std)+'.mat')
                                              ,'_'.join([img_num,'n','{0:03}'.format(nPixels_on_side), 'std', '{0:02}'.format(i_std)+'.mat']))
        print 'Saving params to ',mat_filename 

        pm = sw.superpixels.params                  
        sio.savemat(mat_filename,{'pixel_to_super':sw.seg.cpu, 'count':pm.counts.cpu,
                                  'mu_i': pm.mu_i.cpu, 'mu_s':pm.mu_s.cpu, 
                                  'sigma_s':pm.Sigma_s.cpu, 'sigma_i':pm.Sigma_i.cpu})
        if (not imgs_of_the_same_size):
            del sw


if __name__ == "__main__":  
    import argparse
     
    parser = argparse.ArgumentParser()
    parser.add_argument("-d","--image_direc", help="the directory to work on",
                        nargs='?',
                        const='image',default='image')
    parser.add_argument("-n","--nPixels_on_side", type=int, help="the desired number of pixels on the side of a superpixel",
                        nargs='?',
                        const=10,default=10)
    parser.add_argument("--i_std", type=int, help="std dev for color Gaussians, in [5,40]. A smaller value leads to more irregular superpixels",
                        nargs='?',
                        const=15,default=15)
    
    # If not passed, the value below defaults to False
    parser.add_argument("--imgs_of_the_same_size",
                        help="If passed, it will be assumed that all images in the directory have the same size. This will save some time.",
                        action="store_true")

    
    args = parser.parse_args()    
    main(**args.__dict__)    
    

File: cpp/demo_for_direc.cpp
#include <dirent.h>

#include "Superpixels.h"


static void show_usage(std::string name){
    std::cerr << "Usage of " << name << ":\n"
              << "\t-h, --help \tShow this help message\n"
              << "\t-d, --image_direc \tthe directory to work on\n"
              << "\t-n, --nPixels_on_side \tthe desired number of pixels on the side of a superpixel\n"
              << "\t--i_std \tstd dev for color Gaussians, should be 5<= value <=40. A smaller value leads to more irregular superpixels\n"
              << std::endl;
}


static superpixel_options get_sp_options(int nPixels_in_square_side, int i_std){
    superpixel_options opt;
    opt.nPixels_in_square_side = nPixels_in_square_side; 
    opt.i_std = i_std;

    opt.area = opt.nPixels_in_square_side*opt.nPixels_in_square_side;   
    opt.s_std = opt.nPixels_in_square_side;
    opt.prior_count = 5 * opt.area;
    opt.permute_seg = false;
    opt.calc_cov = true;
    opt.use_hex = true;

    opt.nEMIters = opt.nPixels_in_square_side;
    opt.nInnerIters = 2;
    return opt;
}


// ./Sp_demo_for_direc.py -d <directory_name> -i_std ... -n... --imgs_of_the_same_size 
int main( int argc, char** argv )
{
    
    // get the image filename, nPixels_in_square_side and i_std
    // the defaults
    const char* direc = "image/";
    int nPixels_in_square_side = 15;
    int i_std = 20;
    bool same_size = false;

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if ((arg == "-h") || (arg == "--help")) {
            show_usage(argv[0]);
            return 0;
        } 
        else if ((arg == "-d") || (arg == "--image_direc")) {
            if (i + 1 < argc) { 
                i++;
                direc = argv[i];
            } else {
                std::cerr << "--img_filename option requires one argument." << std::endl;
                return 1;
            }  
        } 
        else if ((arg == "-n") || (arg == "--nPixels_on_side")) {
            if (i + 1 < argc) { 
                i++;
                nPixels_in_square_side = atoi (argv[i]);
                if (nPixels_in_square_side<3) {
                    std::cerr << "--nPixels_in_square_side option requires nPixels_in_square_side >= 3." << std::endl;
                    return 1;
                }
            } else {
                std::cerr << "--nPixels_on_side option requires one argument." << std::endl;
                return 1;
            }  
        }
        else if (arg == "--i_std") {
            if (i + 1 < argc) { 
                i++;
                i_std = atoi (argv[i]);
                
                if (i_std<5 || i_std>40) {
                    std::cerr << "--i_std option requires 5<= value <=40." << std::endl;
                    return 1;
                }
            } else {
                std::cerr << "--i_std option requires a number." << std::endl;
                return 1;
            }  
        }else{  

        }
    }

    DIR *dpdf;
    struct dirent *epdf;

    superpixel_options spoptions = get_sp_options(nPixels_in_square_side,i_std);
    int count = 0;
    //Superpixels sp = Superpixels(0,0,spoptions);
    dpdf = opendir(direc);
    if (dpdf != NULL){

        while (epdf = readdir(dpdf)){     
            String img_name =  epdf->d_name;
              
            String filename = string(direc) + img_name;

            Mat image = imread(filename, CV_LOAD_IMAGE_COLOR);
            if(! image.data )continue;
            
            cout << "Filename: " << filename <<endl;
            Superpixels sp = Superpixels(image.cols, image.rows, spoptions);
            cudaDeviceSynchronize();
            
            //cout << "finish init sp" << endl;
            sp.load_img((unsigned char*)(image.data));
            
            //cout << "finish loading the image" << endl;

            // Part 3: Do the superpixel segmentation
            clock_t start,finish;
            start = clock();
            sp.calc_seg();
            cudaDeviceSynchronize();
            sp.gpu2cpu();
            
            finish = clock();
            cout<< "Segmentation takes " << ((double)(finish-start)/CLOCKS_PER_SEC) << " sec" << endl;

            // Part 4: Save the mean/boundary image 
            cudaError_t err_t = cudaDeviceSynchronize();
            if (err_t){
                std::cerr << "CUDA error after cudaDeviceSynchronize." << std::endl;
                return 0;
            }

            String img_number =  img_name.substr (0, img_name.find(".")); 

            Mat border_img = sp.get_img_overlaid();
            String fname_res_border = "image/result/"+img_number+"_border.png";
            imwrite(fname_res_border, border_img);
            cout << "saving " << fname_res_border << endl;
            
            
            Mat mean_img = sp.get_img_cartoon();
            String fname_res_mean = "image/result/"+img_number+"_mean.png";
            imwrite(fname_res_mean, mean_img);
            cout << "saving " << fname_res_mean << endl << endl;           
            count++;
           
        }
    }
    return 0;
}